rasterio/crs.pyx

```diff
@@ -284,14 +287,18 @@ cdef class CRS:
         CRSError
 
         """
+        if self._geodetic_crs is not None:
+            return self._geodetic_crs if self._geodetic_crs is not False else None
```

Comment:
This will suffice

```suggestion
        if self._geodetic_crs:
            return self._geodetic_crs
```

Reply:
I may be wrong but I think the potential issue would be that the else block could be computationally expensive. The prior code matched the style in pyproj which set the _geodetic_crs to `False` if the conversion failed and the prior if checks cached that by using the False to signify that no conversion was possible. Updating this would cause `GET_GEOG_CS or None` to be called repeatedly

Reply:
How about this: if no geodetic CRS can be determined, which is hopefully rare, we raise `CRSError` and do not cache.

For the user this means that their program will halt if they've made the assumption that there is always a geodetic CRS. But returning `None` to the user has pretty much the same impact: they would always need to check that the value is not None before continuing. And in the case of returning `None` we can't give the user any error message like we would put in an exception.

It all comes down to how we want to use this feature. @AndrewAnnex can you show me a real Python example of how you would use this property in an application? I should have asked for that up front, but I'm a little distracted and distressed this week.

Reply:
Wait, there's more! 😄 

Defining errors out of existence is a good thing and I like to do it. For example, list slicing doesn't raise, it returns empty lists. Thus 

```
for item in [][:-2]:  # last two items of an empty list
    print(item)
```

just prints nothing and there's no error. This kind of expression is very common in a Python program's critical path.

It's a little harder to cleanly define errors away for methods that return a single object.

Reply:
@sgillies at the moment I suppose I don't have any specific lines of code I am aware of off the top of my head where the performance of this function is a expected critical path. I was reminded by some of the prior assumptions for `to_epsg` and `__eq__` though, and since pyproj does cache the property (as per comments from @snowman2), my guess is that the performance was critical elsewhere to that project or it's users, but I didn't look for specific issues/commits related to that. That said, as a user I generally expect accessing properties of objects to be fast (even if there is an expensive initial cache step to pay first).  

The original interest in this function was for some places in the developmentseed morecantile/titiler projects, where users provide query geometries in arbitrary CRSs (projected or geographic coordinates) and checks on the CRS object equality are performed to avoid defining expensive transformation objects/performing expensive transformations. So performance could be important for those projects, but I am speculating somewhat.

I don't think it will be a common occurrence to handle projected CRSs with no geodetic CRS. I attempted to make one using WKT2/proj4 strings and I either would get parsing errors or projections with defined geodetic CRSs (eg `+proj=tmerc +lat_0=0 +lon_0=0 +k=1.0 +x_0=0 +y_0=0 +R=1 +no_defs` still ends up with a geodetic crs as you must specify an ellipsoid or sphere).

I was looking at GDAL's commit history for OSRCloneGeogCS and found some places where null checks are performed, but that seems to be more of a LBYL-pattern exercise. I'd imagine at the C/C++ level, some CRSs are constructed programmatically and it'd be possible to have a null nodes by user error in places, but probably not something rasterio users will run into before other parsing errors occur.

Reply:
@vincentsarago Do you see where this would slot into titiler? Can you link the code?

Reply:
Well for now we rely on pyproj method because the CRS in titiler comes from pyproj https://github.com/developmentseed/morecantile/blob/main/morecantile/models.py#L560-L568

I clearly do not account for the method to return `None` 😭 

I guess Andrew just followed what was in pyproj https://github.com/pyproj4/pyproj/blob/main/pyproj/_crs.pyx#L2677-L2678 

I'm not sure of the intent here but I would be encline to do what pyproj does 😅

Reply:
yes the intent was to match the behavior/functionality of pyproj's `geodetic_crs` function so it'd go against the intent to behave differently than pyproj here. 

At the moment, I don't think it's even possible for pyproj's method to return a None, and I couldn't find any tests in pyproj or gdal to demonstrate that (@snowman2 any examples you know of?). So potentially we would not need to worry about that.

Reply:
@AndrewAnnex @vincentsarago I did some searching, and found that this https://github.com/developmentseed/titiler/blob/0075d1c40b7f4bcbca74a9ae44a7fdfecb9c3718/src/titiler/mosaic/titiler/mosaic/factory.py#L405 is the typical use. To get and write out (e.g. to tileset JSON or STAC) the geographic bounds of a dataset in the context, possibly, of many datasets in a STAC collection. Yeah?

It seems like tileset JSON and STAC already makes the assumption that there is a geodetic CRS for every projected CRS, yes? Unless I've missed something, and there is a strong counter-argument, let's have `geodetic_crs()` to raise `CRSError`. This would prevent you from writing subtly broken tileset JSON or STAC output where `None/null` isn't appropriate or expected, and would let you know if your PROJ install was broken (the CRSError message would say something to that effect).

The caller, not the property itself, would be responsible for handling the CRSError. It does not look like that would be a big deal for, e.g., morecantile. This method in `morecantile.models.py` https://github.com/developmentseed/morecantile/blob/0ab27a724b762764eaa4c320b2cdf219a7d1d06c/morecantile/models.py#L566 would only need to be changed to

```python
    @cached_property
    def rasterio_geographic_crs(self):
        """Return the geographic CRS as a rasterio CRS."""
        try:
            return self.crs.geodetic_crs
        except CRSError:
            log.info("something")
            return None
```

Reply:
@AndrewAnnex yes, I think the first of the two code examples above my comment would be just right.

Reply:
@sgillies I think thats correct. And yeah if a PROJ install is somehow broken (incidentally, I've had a issue lately where occasionally the proj.db file isn't located when using the conda-forge feedstocks, but I hadn't had time to fully diagnose/file a bug report) it'd be good to know that although I'd call that more of an upstream issue. 

Generally I think it's okay to pass the responsibility to the caller, but pyproj made a different choice on that. 

I'll push an update to use option 1